L63-python-eventlet-support.md

```diff
@@ -0,0 +1,108 @@
+Title
+----
+* Author(s): Akrog
+* Approver: a11r
+* Status: Draft
+* Implemented in: python
+* Last updated: 2/21/2020
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+Add an experimental custom IO manager to provide compatibility between the gRPC
```

Comment:
Is the new API/pattern available?  Is there any documentation on it or on what it will look like?

Would it be a possibility to merge this I/O manager and then migrate gevent and eventlet's custom managers at the same time?

Given their similarities I could probably do both, since I've familiarized myself with the differences.

Reply:
The new API is not available yet, but will be available in one or two months. @guantaol Do we have a design doc to share?

If this experimental API got used by third party libraries, like the ones you mentioned below, it would be quite difficult to update all of them. I think it would be better to get it right even with experimental API, if we want people to use it.

I'm sorry that the I/O integration with Core is in a state of limbo. I understand the value of this integration, but we might need a better solution.

Reply:
I don't understand what you mean about "getting it right".  Are you referring about doing it with the new API?

I believe there isn't a problem with those libraries using the proposed code now and then using code that uses the new API.  After all the API surface area with the custom I/O manager is as small as it gets, it's a single method: `init_eventlet`.

So as long as we remain backward compatible (which we would want, just like in the gevent case) then, when we update the eventlet I/O manager to use the new API, everything will be fine and it won't matter how many projects are using it.  Upgrading the gRPC Python library will be seamless.

Reply:
We are planning to have the new EventManager API available by the end of April.  We do not want to accept new code using the existing iomgr APIs at this point, because that will make it harder for us to migrate from the iomgr APIs to the new EventManager API.

Reply:
Is there any code/documentation available on the new API? I can't seen to find the gRFC proposal.

Reply:
I am still working on the new API and will file a gRFC after I make them build on all platforms. The current ETA is end of April.

Reply:
I am disappointed to hear the the gRPC community won't accept a new custom I/O manager considering:

- There is no alternative interface.
- There is no gRFC proposal yet for such a big change.
- There is no communication I can find of not accepting new custom I/O managers in the [meetings doc](https://docs.google.com/document/d/1DTMEbBNmzNbZBh8nOivsnnw3CwUr1Q7WGRe7rNxyHOU), nor in the [Python](https://github.com/grpc/grpc/tree/bbdc86df1b6de9fc59382f3197f3e4df6f19a0f7/src/python), [grpcio](https://github.com/grpc/grpc/tree/bbdc86df1b6de9fc59382f3197f3e4df6f19a0f7/src/python/grpcio), or [experimental](https://github.com/grpc/grpc/tree/bbdc86df1b6de9fc59382f3197f3e4df6f19a0f7/src/python/grpcio/grpc/experimental) directories.
- I have offered to port both the gevent and eventlet managers to the new interface once it's available.
- The port to the new interface will be backward compatible and won't affect any existing integration.
- Testing failure ratio on Linux has been determined to be ~1‰ (6 failures out of 6200 test runs).